File: dart/test/decode_helpers_test.dart
import 'package:flat_buffers/flex_buffers.dart' as flex;
import 'package:ob_dump_reader/ob_dump_reader.dart';
import 'package:test/test.dart';

void main() {
  test('bytesToHex encodes raw bytes as lowercase hex', () {
    expect(bytesToHex([0xde, 0xad, 0xbe, 0xef]), 'deadbeef');
    expect(bytesToHex([]), '');
  });

  test('bytesToUuidString groups 16 bytes into canonical form', () {
    final uuidBytes = [
      0x55, 0x0e, 0x84, 0x00, 0xe2, 0x9b, 0x41, 0xd4,
      0xa7, 0x16, 0x44, 0x66, 0x55, 0x44, 0x00, 0x00,
    ];
    expect(bytesToUuidString(uuidBytes), '550e8400-e29b-41d4-a716-446655440000');
  });

  test('bytesToUuidString falls back to plain hex for non-16-byte input', () {
    expect(bytesToUuidString([0xde, 0xad]), 'dead');
  });

  test('tryParseJsonString parses valid JSON', () {
    expect(tryParseJsonString('{"a":1,"b":[true,null]}'), {
      'a': 1,
      'b': [true, null],
    });
    expect(tryParseJsonString('[1,2,3]'), [1, 2, 3]);
  });

  test('tryParseJsonString falls back to the original string on invalid JSON', () {
    const jsSource = 'function() { return 1; }';
    expect(tryParseJsonString(jsSource), jsSource);
  });

  test('decodeFlex decodes a FlexBuffers map to a Dart Map', () {
    final buffer = flex.Builder.buildFromObject({'a': 1, 'b': true, 'c': null});
    final bytes = buffer.asUint8List();

    expect(decodeFlex(bytes), {'a': 1, 'b': true, 'c': null});
  });

  test('decodeFlex decodes a FlexBuffers vector root to a Dart List', () {
    final buffer = flex.Builder.buildFromObject([1, 2, 'three']);
    final bytes = buffer.asUint8List();

    expect(decodeFlex(bytes), [1, 2, 'three']);
  });
}

File: js/test/decodeHelpers.test.ts
import { encode } from "flatbuffers/js/flexbuffers.js";
import { describe, expect, it } from "vitest";

import {
  bytesToHex,
  bytesToUuidString,
  decodeFlex,
  readUint32be,
  tryParseJsonString,
  uint32be,
} from "../src/decodeHelpers.js";

describe("readUint32be / uint32be", () => {
  it("round-trips a big-endian uint32", () => {
    const value = 0x01020304;
    expect(readUint32be(Buffer.from(uint32be(value)), 0)).toBe(value);
  });

  it("reads at a non-zero offset", () => {
    const buf = Buffer.concat([Buffer.from([0xff]), Buffer.from(uint32be(42))]);
    expect(readUint32be(buf, 1)).toBe(42);
  });
});

describe("bytesToHex", () => {
  it("hex-encodes raw bytes", () => {
    expect(bytesToHex(Buffer.from([0xde, 0xad, 0xbe, 0xef]))).toBe("deadbeef");
  });
});

describe("bytesToUuidString", () => {
  it("formats 16 bytes as a canonical UUID", () => {
    const bytes = Buffer.from(
      "0123456789abcdef0123456789abcdef",
      "hex",
    ).subarray(0, 16);
    expect(bytesToUuidString(bytes)).toBe(
      "01234567-89ab-cdef-0123-456789abcdef",
    );
  });

  it("falls back to plain hex for non-16-byte input", () => {
    expect(bytesToUuidString(Buffer.from([1, 2, 3]))).toBe("010203");
  });
});

describe("tryParseJsonString", () => {
  it("parses valid JSON", () => {
    expect(tryParseJsonString('{"a":1}')).toEqual({ a: 1 });
  });

  it("returns the original string when parsing fails", () => {
    expect(tryParseJsonString("not json")).toBe("not json");
  });
});

describe("decodeFlex", () => {
  it("decodes a FlexBuffers map/vector/scalar mix into a plain JS value", () => {
    const bytes = encode({ a: 1, b: ["x", "y"], c: true });
    expect(decodeFlex(bytes)).toEqual({ a: 1, b: ["x", "y"], c: true });
  });

  it("decodes correctly when the bytes are a view into a larger pooled buffer", () => {
    // Node's small Buffer.from/allocations are frequently slices of a
    // shared underlying ArrayBuffer (byteOffset != 0) — the exact case
    // `decodeFlex`'s ArrayBuffer conversion has to handle correctly.
    const encoded = Buffer.from(encode({ n: 42 }));
    const padded = Buffer.concat([Buffer.from([0xff, 0xff, 0xff]), encoded]);
    const view = padded.subarray(3);
    expect(view.byteOffset).toBeGreaterThan(0);
    expect(decodeFlex(view)).toEqual({ n: 42 });
  });
});
